libfwupdplugin/fu-progress.c

```diff
@@ -0,0 +1,714 @@
+ *
+ * Since: 1.7.0
+ **/
+static gboolean
+fu_progress_get_profile(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_PROGRESS(self), FALSE);
+	return priv->profile;
+}
+
+/**
+ * fu_progress_reset:
+ * @self: A #FuProgress
+ *
+ * Resets the #FuProgress object to unset
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_reset(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_PROGRESS(self));
+
+	/* reset values */
+	priv->step_max = 0;
+	priv->step_now = 0;
+	priv->percentage = 0;
+
+	/* only use the timer if profiling; it's expensive */
+	if (priv->profile)
+		g_timer_start(priv->timer);
+
+	/* disconnect client */
+	if (priv->percentage_child_id != 0) {
+		g_signal_handler_disconnect(priv->child, priv->percentage_child_id);
+		priv->percentage_child_id = 0;
+	}
+
+	/* unref child */
+	g_clear_object(&priv->child);
+
+	/* no more step data */
+	g_clear_pointer(&priv->step_data, g_free);
+	g_clear_pointer(&priv->step_profile, g_free);
+}
+
+/**
+ * fu_progress_set_steps_full:
+ * @self: A #FuProgress
+ * @id: the code location
+ * @step_max: The number of sub-tasks in this progress, can be 0
+ *
+ * Sets the number of sub-tasks, i.e. how many times the fu_progress_step_done()
+ * function will be called in the loop.
+ *
+ * The function will immediately return when the number of step_max is 0
+ * or if fu_progress_set_enabled(FALSE) was previously called.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_set_steps_full(FuProgress *self, const gchar *id, guint step_max)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_PROGRESS(self));
+
+	/* nothing to do for 0 step_max */
+	if (step_max == 0)
+		return;
+
+	/* did we call done on a self that did not have a size set? */
+	if (priv->step_max != 0) {
+		g_autoptr(GString) str = g_string_new(NULL);
+		fu_progress_build_parent_chain(self, str, 0);
+		g_warning("step_max already set to %u, can't set %u! [%s]: %s",
+			  priv->step_max,
+			  step_max,
+			  id,
+			  str->str);
+		return;
+	}
+
+	/* set id */
+	g_free(priv->id);
+	priv->id = g_strdup(id);
+
+	/* only use the timer if profiling; it's expensive */
+	if (priv->profile)
+		g_timer_start(priv->timer);
+
+	/* imply reset */
+	fu_progress_reset(self);
+
+	/* set step_max */
+	priv->step_max = step_max;
+
+	/* global share just got smaller */
+	priv->global_share /= step_max;
+}
+
+/**
+ * fu_progress_set_custom_steps_full:
+ * @self: A #FuProgress
+ * @id: the code location
+ * @value: A step weighting variable argument array
+ *
+ * This sets the step weighting, which you will want to do if one action
+ * will take a bigger chunk of time than another.
+ *
+ * All the values must add up to 100, and the list must end with -1.
+ * Do not use this function directly, instead use the fu_progress_set_custom_steps() macro.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_set_custom_steps_full(FuProgress *self, const gchar *id, guint value, ...)
```

Comment:
I don't know about locking into this API of steps are percent of the device update vs time of the device update.  Maybe allow funcs for both to determine later?

Reply:
So, I'm kinda hoping that percentage complete and time-complete will be the same thing, at least if everything uses FuProgress like I'm imagining. At least when you use FWUPD_PROGRESS_PROFILE=1 we use the _time_ between each step to suggest the `fu_progress_set_custom_steps()` values.

Reply:
OK..